feat: TrinoSQLValidator

giftig opened the pull request:
<!---
Please write the PR title following the conventions at https://www.conventionalcommits.org/en/v1.0.0/
Example:
fix(dashboard): load charts correctly
-->
feat: Trino SQL Validator

### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->

Add an SQL validator for Trino, based on the existing Presto one but catching the appropriate exception type. Extend presto SQL validator in order to do this.

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
<!--- Skip this if not applicable -->

### TESTING INSTRUCTIONS
<!--- Required! What steps can be taken to manually verify the changes? -->
- Add a trino database
- Write a query in SQL Lab
- Observe on the fly SQL validation

### ADDITIONAL INFORMATION
<!--- Check any relevant boxes with "x" -->
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->
- [x] Has associated issue: https://github.com/apache/superset/discussions/23851
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [x] Introduces new feature or API
- [ ] Removes existing feature or API

giftig committed:
feat: TrinoSQLValidator

rusackas commented:
Approving CI and tagging 
@cccs-tom
@villebro
@sujiplr
for review and/or input :)

giftig commented:
Fixed the pylint errors; looks like I missed an import when porting the fix from our fork to the base project, and I've added pylint exceptions for local imports. I didn't make the original change but it looks like the `DatabaseError` was local in the presto class because it had been done locally in another place there, so I'm assuming this was necessary.

The mysql test error looks like a flaky test, don't think it's anything to do with my change.

cccs-tom commented on superset/sql_validators/trino.py:
```diff
@@ -0,0 +1,31 @@
+
+from typing import Type
+
+from superset.sql_validators.presto_db import PrestoDBSQLValidator
```
Unfortunately I haven't had a chance to test this (and I'm on a task unrelated to Trino, so it'll be a while before I have a chance to test), but I have some real concerns over re-using the PrestoDB Validator for this. First off, Trino was forked from Presto several years ago now and the projects continue to diverge with each release. But more specifically, I was looking at that PrestoDBSQLValidator class and it does a bunch of error handling that assumes that the underlying library is PyHive. PyHive is no longer the recommended way to connect Superset to Trino and it hasn't been for quite some time. Imo, this implementation should take that into consideration and it should be independent from PrestoDB / PyHive.

giftig replied:
We've been using this in our fork for the past year or so I believe, just to give us some sort of validation for trino queries. It's true I've occasionally seen some errors logged e.g. a missing `poll` attribute on `Cursor` which suggests they're not fully compatible here, and have to fudge the error type does flag the issue. We've found that it's better than nothing, but perhaps not reliable enough to be a useful contribution.

We're likely to go the route of highly customised validation errors and warnings on the validator if we do write a more comprehensive version, but if we take inspiration from what the presto and postgres validators are looking for maybe we can build that in such a way to be able to contribute something useful here and add anything more custom on top.

That said it sounds like trino isn't a big priority in this project  at the moment, and the SQL validators aren't a big focus either, given there have only ever been two of them. Hopefully we can contribute something more useful here once we've iterated it a bit, though.
